Add support for huge pages in SMPI

Ezibenroc opened the pull request:
This adds an option to `smpirun` asking to use huge pages for the calls to `SMPI_SHARED_MALLOC` and `SMPI_PARTIAL_SHARED_MALLOC`.

To use it:
```bash
mkdir /home/huge && sudo mount none /home/huge -t hugetlbfs -o rw,mode=0777
sudo sh -c "/bin/echo 1 > /proc/sys/vm/nr_hugepages"
smpirun --cfg=smpi/shared-malloc-hugepage:/home/huge --cfg=smpi/shared-malloc-blocksize:2097152 [...]
```

For very large allocations, the page table grows very large and the kernel takes a lot of time working on this table, stalling the simulation.
For instance, without huge pages, HPL with a matrix of size 6e5 has a page table of 5.5GB and the simulation drops below 40% of CPU utilization.
Using huge pages fix this. Of course this only shifts the problem further, but this is an easy fix.

Currently, the size of a huge page is hard coded to be 2 MB, as I did not find a standard way to get it. Maybe we could just ask the user to provide it?

mquinson commented:
Could you please add a test for it? It can be any existing test with the new flag passed, I guess.
If you don't want to mess with the cmake madness, just tell me the test you want to change and I'll take care of this.

Many thanks for this great feature!
Mt

mquinson commented on src/smpi/smpi_shared.cpp:
```diff
@@ -125,6 +127,7 @@ std::map<void*, shared_metadata_t> allocs_metadata;
 xbt_dict_t calls = nullptr;           /* Allocated on first use */
 #ifndef WIN32
 static int smpi_shared_malloc_bogusfile           = -1;
+static int smpi_shared_malloc_bogusfile_huge_page  = -1;
```
Why do you need another global variable `smpi_shared_malloc_bogusfile_huge_page`? Can't you use  the `smpi_shared_malloc_bogusfile` variable in any case, regardless of whether we use huge pages or regular ones?

Ezibenroc replied:
When huge pages are used, there are still some "classical" calls to mmap without `MAP_HUGETLB` for areas smaller than a huge page (the start and the end of the blocks).
Apparently, you cannot call mmap without `MAP_HUGETLB` on a file opened in a `hugetlbfs` mount point. So you need two files in this case.

mquinson replied:
Damn, Ok. I think you should add a little comment about it, maybe.

mquinson commented on src/smpi/smpi_shared.cpp:
```diff
@@ -218,23 +223,63 @@ static void *smpi_shared_malloc_local(size_t size, const char *file, int line)
-  xbt_assert(smpi_shared_malloc_blocksize % PAGE_SIZE == 0, "The block size of shared malloc should be a multiple of the page size.");
+  char *huge_page_mount_point = xbt_cfg_get_string("smpi/shared-malloc-hugepage");
+  const bool use_huge_page = huge_page_mount_point[0] != '\0';
+#ifndef MAP_HUGETLB
```
Please don't use #ifndef, because it makes it hard to detect whether you made a typo in the name (or removed a needed header file). Instead, use `#if !MAP_HUGETLB` or similar. With this, the `-Wundef` that we are using can detect the previous errors.

mquinson replied:
Ok, I understand now that this is a standard mmap flag, defined in system headers. Sorry for the mistake. Please explain it with something like (to be used each time you have such a test):
```
 #ifndef MAP_HUGETLB /* If the system header don't define that mmap flag */
```

mquinson commented on src/smpi/smpi_shared.cpp:
```diff
@@ -218,23 +223,63 @@ static void *smpi_shared_malloc_local(size_t size, const char *file, int line)
+  char *huge_page_mount_point = xbt_cfg_get_string("smpi/shared-malloc-hugepage");
+  const bool use_huge_page = huge_page_mount_point[0] != '\0';
+#ifndef MAP_HUGETLB
+    xbt_assert(!use_huge_page, "Huge pages are not available on your system.");
```
Be more explicit for the user: add "don't use the smpi/shared-malloc-hugepage option"

mquinson commented on src/smpi/smpi_shared.cpp:
```diff
@@ -218,23 +223,63 @@ static void *smpi_shared_malloc_local(size_t size, const char *file, int line)
+  smpi_shared_malloc_blocksize = static_cast<unsigned long>(xbt_cfg_get_double("smpi/shared-malloc-blocksize"));
+  void *mem, *allocated_ptr;
+  size_t allocated_size;
+#ifdef MAP_HUGETLB
```
You can factorize this if block by defining use_huge_page to false when not MAP_HUGETLB

mquinson commented on src/smpi/smpi_shared.cpp:
```diff
@@ -218,23 +223,63 @@ static void *smpi_shared_malloc_local(size_t size, const char *file, int line)
+  xbt_assert(allocated_ptr != MAP_FAILED, "Failed to allocate %zuMiB of memory. Run \"sysctl vm.overcommit_memory=1\" as root "
                                 "to allow big allocations.\n",
              size >> 20);
+#ifdef MAP_HUGETLB
```
same factorization needed here

mquinson commented:
Also, you need to document the option (and how to use it) in doc/doxygen/options.doc, please.
Thanks,

Ezibenroc commented:
Good point for the factorization, it is more readable now.

I added the documentation for the huge page inside the section of shared malloc and not in a new section, because I think that it is more relevant here. I also added the documentation for the partial shared malloc that was missing.

mquinson commented on doc/doxygen/options.doc:
```diff
@@ -1115,6 +1115,35 @@ only consume 1 MiB in memory.
+
+Then, you can pass the option --cfg=smpi/shared-malloc-hugepage:/home/huge
+to smpirun.
+
```
Very good, thanks. You could improve it by saying that "echo 1" is the minimal size that we need but that any larger value is ok, but that's not mandatory. It's really good as is. Instructive and succint.

mquinson commented:
We still don't know how to test it, though. We create a hugetlbfs mountpoint to a specific location (like ~/hugepages) on several jenkins hosts, and have the cmake using it if it exists.

Or we try to have cmake using /dev/hugepages by default, which rights should be changed on the  jenkins hosts.

@degomme, what do you think?

Ezibenroc commented:
Correct me if I missed something, but this feature is still not tested. We should do something.

mquinson commented:
I agree we should test it. But how? We could change the permission on /dev/hugepages on some hosts on Jenkins, and have a test trying to use it if possible? This could do the trick. 

I'm reopening+retitling the issue until this is fixed.

Thanks for the poke.
